marshmallow_jsonapi/fields.py

```diff
@@ -115,24 +120,28 @@ def validate_data_object(self, data):
 
         if errors:
             raise ValidationError(errors)
+        return data.get('id')
 
-    def _deserialize(self, value, attr, obj):
-        if 'data' not in value:
-            raise ValidationError('Must include a `data` key')
+    def deserialize(self, value, attr=None, data=None):
+        """Deserialize ``value``.
 
-        data = value.get('data')
-        if data is None or value['data'] == []:
-            return data
+        :raise ValidationError: If the value is not type `dict`, if the
+            value does not contain a `data` key, and if the value is
+            required but unspecified.
+        """
+        if not isinstance(value, dict) or 'data' not in value:
+            raise ValidationError('Must include a `data` key')
+        return super(Relationship, self).deserialize(value['data'], attr, data)
 
+    def _deserialize(self, value, attr, obj):
         if self.many:
-            for item in data:
-                self.validate_data_object(item)
-        else:
-            self.validate_data_object(data)
+            if not isinstance(value, list):
+                raise ValidationError('Relationship is list-like')
+            return [self.extract_value(item) for item in value]
 
-        if self.many:
-            return [item.get('id') for item in data]
-        return data.get('id')
+        if isinstance(value, list):
```

Comment:
You can use `marshmallow.utils.is_collection`, which will allow for custom collection classes.